### PR TITLE
feat: Implement `MemoryUsage` for `Store`

### PR DESCRIPTION
**Pull request by Hywan**
# Description

This patch implements `loupe::MemoryUsage` for `wasmer::Store`.

# Review

- [x] Add a short description of the the change to the CHANGELOG.md file

### CODE REVIEW

File: lib/api/Cargo.toml
Comment (syrusakbary) [inline lib/api/Cargo.toml:31]: Lets publish the crate before merging the PR.
Having this comment as a reminder :)

File: lib/compiler/src/target.rs
Comment (syrusakbary) [inline lib/compiler/src/target.rs:166]: Why ignoring this?

File: lib/compiler/src/target.rs
Comment (syrusakbary) [inline lib/compiler/src/target.rs:168]: Why ignoring this?

File: lib/engine/Cargo.toml
Comment (syrusakbary) [inline lib/engine/Cargo.toml:29]: Fix me

File: lib/middlewares/Cargo.toml
Comment (syrusakbary) [inline lib/middlewares/Cargo.toml:18]: Fix me

File: lib/vm/Cargo.toml
Comment (syrusakbary) [inline lib/vm/Cargo.toml:25]: Fix me

File: lib/wasmer-types/Cargo.toml
Comment (syrusakbary) [inline lib/wasmer-types/Cargo.toml:17]: Fix me

File: tests/lib/engine-dummy/Cargo.toml
Comment (syrusakbary) [inline tests/lib/engine-dummy/Cargo.toml:19]: Fix me

File: lib/compiler-cranelift/Cargo.toml
Comment (syrusakbary) [inline lib/compiler-cranelift/Cargo.toml:29]: Fix me

File: lib/compiler-llvm/Cargo.toml
Comment (syrusakbary) [inline lib/compiler-llvm/Cargo.toml:26]: Fix me

File: lib/compiler/Cargo.toml
Comment (syrusakbary) [inline lib/compiler/Cargo.toml:25]: Fix me

File: lib/engine-jit/Cargo.toml
Comment (syrusakbary) [inline lib/engine-jit/Cargo.toml:25]: Fix me

File: lib/engine-native/Cargo.toml
Comment (syrusakbary) [inline lib/engine-native/Cargo.toml:28]: Fix me

File: lib/engine-object-file/Cargo.toml
Comment (syrusakbary) [inline lib/engine-object-file/Cargo.toml:27]: Fix me

File: lib/api/Cargo.toml
Comment (Hywan) [reply lib/api/Cargo.toml:31]: Absolutely. It's a WIP, I'm modifying `loupe` quite a lot for the moment. As soon as we finish this project, I will release `loupe` and update this PR.

File: lib/compiler/src/target.rs
Comment (Hywan) [reply lib/compiler/src/target.rs:166]: Because we don't own this type. We can't implement `MemoryUsage` on this type for the moment. We may want to implement `MemoryUsage` manually on `Target` and try a custom computation though.

File: lib/compiler/src/target.rs
Comment (Hywan) [reply lib/compiler/src/target.rs:168]: Because we don't own this type. We can't implement `MemoryUsage` on this type for the moment. We may want to implement `MemoryUsage` manually on `Target` and try a custom computation though.

**Comment by Hywan** [comment]
bors try

**Comment by Hywan** [comment]
bors try

File: examples/tunables_limit_memory.rs
Comment (MarkMcCaskey) [inline examples/tunables_limit_memory.rs:141]: It's a bit surprising to me that this isn't something the macro can generate. I'm not sure how common of a base case it is, maybe it isn't worth it.

File: lib/compiler/src/target.rs
Comment (MarkMcCaskey) [inline lib/compiler/src/target.rs:170]: This might be the same as the thing I mentioned above, not high priority and easy to add later, but might be able to do something like

```
#[loupe(skip_all)]
pub struct Target {
...
}
```

To just get the local size and not any of the heap size stuff

**Review by MarkMcCaskey — APPROVED** [review]
Looks good to me other than the fact that we shouldn't merge this upstream until we publish `loupe` to crates.io 👍

File: examples/tunables_limit_memory.rs
Comment (Hywan) [reply examples/tunables_limit_memory.rs:141]: It's because of the generic here. I need to dig into `loupe-derive` to see what's going on.

File: lib/compiler/src/target.rs
Comment (Hywan) [reply lib/compiler/src/target.rs:170]: I've created https://github.com/wasmerio/loupe/issues/12.

**Comment by Hywan** [comment]
bors try

**Review by syrusakbary — APPROVED** [review]

File: examples/tunables_limit_memory.rs
Comment (Hywan) [reply examples/tunables_limit_memory.rs:141]: Fixed by https://github.com/wasmerio/loupe/pull/14. I'll update this file in #2200 probably.

**Pull request merged**
